src/day23.scala

```diff
@@ -0,0 +1,151 @@
+    amphipodPositions.flatMap((amphipod, positions) => positions.map(position => position -> amphipod))
+
+  def move(amphipod: Amphipod): Set[(Situation, Energy)] =
+    for
+      from <- amphipodPositions(amphipod)
+      stop <- nextStops(amphipod, from)
+      path = getPath(from, stop)
+      if path.forall(isEmpty)
+    yield
+      val newPositions = amphipodPositions(amphipod) - from + stop
+      val energy = path.size * amphipod.energy
+      (copy(amphipodPositions + (amphipod -> newPositions)), energy)
+  
+  def isFinal =
+    amphipodPositions.forall { (amphipod, positions) =>
+      positions.forall(p => p.x == amphipod.destinationX)
+    }
+
+  private def getDestination(amphipod: Amphipod): Option[Position] =
+    val x = amphipod.destinationX
+    @tailrec
+    def tailApply(y: Int): Option[Position] =
+      positionToAmphipod.get(Position(x, y)) match
+        case Some(other) if other == amphipod => if y == 2 then None else tailApply(y - 1)
+        case Some(_) => None
+        case None => Some(Position(x, y))
+    tailApply(roomSize + 1)
+  
+  private def nextStops(amphipod: Amphipod, from: Position): Seq[Position] =
+    val destination = getDestination(amphipod)
+    from match
+      case Position(x, y) if x == amphipod.destinationX =>
+        destination match
+          case Some(destination) => Seq.empty
+          case None => hallwayStops
+      case Position(_, 1) => destination.toSeq
+      case _ => hallwayStops ++ destination
+
+  private def getPath(start: Position, stop: Position): Seq[Position] =
+    val hallway = 
+      if start.x < stop.x
+      then (start.x + 1).to(stop.x).map(Position(_, 1))
+      else (start.x - 1).to(stop.x, step = -1).map(Position(_, 1))
+    val startRoom = (start.y - 1).to(1, step = - 1).map(Position(start.x, _))
+    val stopRoom = 2.to(stop.y).map(Position(stop.x, _))
+    startRoom ++ hallway ++ stopRoom
+
+  private def isEmpty(position: Position) =
+    !positionToAmphipod.contains(position)
+
+val hallwayStops: Seq[Position] = Seq(
+  Position(1, 1),
+  Position(2, 1),
+  Position(4, 1),
+  Position(6, 1),
+  Position(8, 1),
+  Position(10, 1),
+  Position(11, 1)
+)
+
+@tailrec
+def findCheapestSolution(currentSituations: Map[Situation, Energy], movingAmphipod: Amphipod): Energy =
+  println(s"moving $movingAmphipod")
+  val nextSituations = for
+    (currentSituation, currentEnergy) <- currentSituations.toSeq
+    (nextSituation, energy) <- currentSituation.move(movingAmphipod)
+  yield (nextSituation, currentEnergy + energy)
+  
+  nextSituations.find((situation, energy) => situation.isFinal) match
+    case Some((_, energy)) => energy
+    case None =>
+      val allSituations = nextSituations.foldLeft(currentSituations) { case (acc, (situation, energy)) => 
+        val minEnergy = math.min(acc.getOrElse(situation, Int.MaxValue), energy)
+        acc.updated(situation, minEnergy)
+      }
+      if allSituations != currentSituations then 
```

Comment:
This is probably quite expensive (it's an equality test between unordered collections). Isn't there any constant-time way you can detect whether you should go to `Amphipod.A`? Couldn't you do that always when `movingAmphipod == D`, for example?